File: app/src/main/java/com/soundcloud/android/discovery/recommendedplaylists/RecommendedPlaylistsEntity.java
package com.soundcloud.android.discovery.recommendedplaylists;

import com.google.auto.value.AutoValue;
import com.soundcloud.android.model.Urn;
import com.soundcloud.java.optional.Optional;

import java.util.Collections;
import java.util.List;

@AutoValue
public abstract class RecommendedPlaylistsEntity {

    public abstract Long localId();
    public abstract String key();
    public abstract String displayName();
    public abstract Optional<String> artworkUrl();
    public abstract Optional<Urn> queryUrn();
    public abstract List<Urn> playlistUrns();

    public static RecommendedPlaylistsEntity create(long localId, String key, String displayName, Optional<String> artworkUrl, Optional<Urn> queryUrn, List<Urn> playlists) {
        return new AutoValue_RecommendedPlaylistsEntity(localId, key, displayName, artworkUrl, queryUrn, playlists);
    }

    public RecommendedPlaylistsEntity copyWithPlaylistUrns(List<Urn> playlists) {
        return new AutoValue_RecommendedPlaylistsEntity(localId(), key(), displayName(), artworkUrl(), queryUrn(), playlists);
    }

    public static RecommendedPlaylistsEntity create(long localId, String key, String displayName, Optional<String> artworkUrl, Optional<Urn> queryUrn) {
        return new AutoValue_RecommendedPlaylistsEntity(localId, key, displayName, artworkUrl, queryUrn, Collections.emptyList());
    }
}

File: app/src/main/java/com/soundcloud/android/olddiscovery/recommendedplaylists/RecommendedPlaylistsEntity.kt
package com.soundcloud.android.olddiscovery.recommendedplaylists

import com.soundcloud.android.model.Urn
import com.soundcloud.java.optional.Optional

@Suppress("LongParameterList")
data class RecommendedPlaylistsEntity(
        val localId: Long,
        val key: String,
        val displayName: String,
        val artworkUrl: String?,
        val queryUrn: Urn?,
        val playlistUrns: List<Urn> = arrayListOf()
) {
    companion object {
        @JvmStatic
        fun create(localId: Long, key: String, displayName: String, artworkUrl: Optional<String>, queryUrn: Optional<Urn>, playlistUrns: List<Urn>): RecommendedPlaylistsEntity {
            return RecommendedPlaylistsEntity(localId, key, displayName, artworkUrl.orNull(), queryUrn.orNull(), playlistUrns)
        }

        @JvmStatic
        fun create(localId: Long, key: String, displayName: String, artworkUrl: Optional<String>, queryUrn: Optional<Urn>): RecommendedPlaylistsEntity {
            return RecommendedPlaylistsEntity(localId, key, displayName, artworkUrl.orNull(), queryUrn.orNull())
        }
    }

    fun getOptionalArtworkUrl(): Optional<String> = Optional.fromNullable(artworkUrl)
    fun getOptionalQueryUrn(): Optional<Urn> = Optional.fromNullable(queryUrn)

    fun copyWithPlaylistUrns(urns: List<Urn>) = copy(playlistUrns = urns)
}
